onnx/onnx.in.proto

```diff
@@ -626,6 +626,90 @@ message TensorShapeProto {
   repeated Dimension dim = 1;
 }
 
+// Sequences and Maps are made of TypeProto elements, which can either be
+// tensors, maps, or sequences.
+message SequenceMapElement {
+  enum DataType {
+    TENSOR = 1;
+    SPARSE_TENSOR = 2;
+    SEQUENCE = 3;
+    MAP = 4;
+  }
+  // The data type of the element.
+  // This field MUST have a valid SequenceMapElement.DataType value
+  optional int32 elem_type = 1;
+
+  optional TensorProto tensor_value = 2;
+  optional SparseTensorProto sparse_tensor_value = 3;
+  optional SequenceProto sequence_value = 4;
+  optional MapProto map_value = 5;
+}
```

Comment:
I'm seeing this as trying to make a complete set of data values representing any kind of sequences/maps.

Can you give some real use cases please? say, some traditional ml models. I guess, not all the compositions (also recursive) of different values (sequences/maps based on tensors/sequences/maps again) are that useful.

Reply:
> I'm seeing this as trying to make a complete set of data values representing any kind of sequences/maps.

Yes, this was the intention as detailed in the spec requirements for sequences and maps. I suppose the use case would be multimodal data input, specifically using a data structure analogous to a pythonic list (sequence) to contain tensor data of different shapes and dictionary data (maps) as a single input to a model.

i.e. [ [ np.array([1, 2, 3]), np.array([2, 3, 4, 5]) ], {1:A, 2:B, 3:C} ]  as model input.

I agree that there's not a lot of use cases for it, but I did think it was the best way to represent the written spec. (Unless we determine that sequences / maps cannot recursively contain themselves).

Reply:
> I'm seeing this as trying to make a complete set of data values representing any kind of sequences/maps.

Yes, defining the data type's complete possibilities is the thought process here. I don't see real-life use cases of these recursively-defined data values. There are comparative usefulness levels for multimodal data, i.e. a map with individual maps as it's own values seems less relevant than a sequence containing a sequence. Any more than 2 dimensions of recursively defining themselves don't seem to be useful as input for the current state-of-the-art in NN models.

Reply:
However, the other side of the coin is that some frameworks do allow input of this form to be expressed, even if they're not being currently used. The alternative in onnx.in.proto would be to create a similar format to key representation in MapProto or data representation in TensorProto with a data_type variable and optional input entries (Tensor, SparseTensor, Map, Sequence).

Reply:
Yes, it is complete. But the completeness does not cost us anything (it doesn't add to the complexity in any way, right?). This seems preferable to alternative restricted versions that allow only 1 or 2 levels of nesting, because sooner or later we might find we need to use 3 levels of nesting, etc. As I mention above, the drawback/complexity is only in bumping IR_VERSION, so I think may be defining a different proto file will solve that problem.